src/server/session.ts

```diff
@@ -2121,26 +2211,73 @@ namespace ts.server {
+
             if (currentFileOnly) {
                 Debug.assertIsDefined(args.file);
                 const { file, project } = this.getFileAndProject(args as protocol.FileRequestArgs);
-                return [{ project, result: project.getLanguageService().getNavigateToItems(searchValue, maxResultCount, file) }];
-            }
-            else if (!args.file && !projectFileName) {
-                return combineProjectOutputFromEveryProject(
-                    this.projectService,
-                    project => project.getLanguageService().getNavigateToItems(searchValue, maxResultCount, /*filename*/ undefined, /*excludeDts*/ project.isNonTsProject()),
-                    navigateToItemIsEqualTo);
-            }
-            const fileArgs = args as protocol.FileRequestArgs;
-            return combineProjectOutputWhileOpeningReferencedProjects<NavigateToItem>(
-                this.getProjects(fileArgs),
-                this.getDefaultProject(fileArgs),
-                project => project.getLanguageService().getNavigateToItems(searchValue, maxResultCount, /*fileName*/ undefined, /*excludeDts*/ project.isNonTsProject()),
-                documentSpanLocation,
-                navigateToItemIsEqualTo);
+                return [{ project, navigateToItems: project.getLanguageService().getNavigateToItems(searchValue, maxResultCount, file) }];
+            }
+
+            const outputs: ProjectNavigateToItems[] = [];
+
+            // This is effectively a hashset with `name` as the custom hash and `navigateToItemIsEqualTo` as the custom equals.
+            // `name` is a very cheap hash function, but we could incorporate other properties to reduce collisions.
+            const seenItems = new Map<string, NavigateToItem[]>(); // name to items with that name
+
+            if (!args.file && !projectFileName) {
+                // VS Code's `Go to symbol in workspaces` sends request like this by default.
+                // There's a setting to have it send a file name (reverting to older behavior).
+
+                // TODO (https://github.com/microsoft/TypeScript/issues/47839)
+                // This appears to have been intended to search all projects but, in practice, it seems to only search
+                // those that are downstream from already-loaded projects.
+                // Filtering by !isSourceOfProjectReferenceRedirect is new, but seems appropriate and consistent with
+                // the case below.
+                this.projectService.loadAncestorProjectTree();
+                this.projectService.forEachEnabledProject(project => addItemsForProject(project));
+            }
+            else {
+                // VS's `Go to symbol` sends requests with just a project and doesn't want cascading since it will
+                // send a separate request for each project of interest
+
+                // TODO (https://github.com/microsoft/TypeScript/issues/47839)

```

Comment:
I basically need Sheetal's input to resolve these, but I think the current behavior is acceptably similar to the current behavior.